pkg/planner/core/memtable_infoschema_extractor.go

```diff
@@ -0,0 +1,493 @@
+}
+
+// ListSchemasAndTables lists related tables and their corresponding schemas from predicate.
+// If there is no error, returning schema slice and table slice are guaranteed to have the same length.
+func (e *InfoSchemaStatisticsExtractor) ListSchemasAndTables(
+	ctx context.Context,
+	is infoschema.InfoSchema,
+) ([]model.CIStr, []*model.TableInfo, error) {
+	schemas := e.listSchemas(is, _tableSchema)
+	tableNames := e.getSchemaObjectNames(_tableName)
+	if len(tableNames) > 0 {
+		return findTableAndSchemaByName(ctx, is, schemas, tableNames)
+	}
+	return listTablesForEachSchema(ctx, is, schemas)
+}
+
+// ListTables lists related tables from predicate.
+// If no table is found in predicate, it return all tables.
+func (e *InfoSchemaStatisticsExtractor) ListTables(
+	ctx context.Context,
+	is infoschema.InfoSchema,
+	schema model.CIStr,
+) ([]*model.TableInfo, error) {
+	tableNames := e.getSchemaObjectNames(_tableName)
+	if len(tableNames) == 0 {
+		return is.SchemaTableInfos(ctx, schema)
+	}
+
+	tables := make(map[int64]*model.TableInfo, 8)
+	err := findNameAndAppendToTableMap(ctx, is, schema, tableNames, tables)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return maps.Values(tables), nil
+}
+
+// InfoSchemaSchemataExtractor is the predicate extractor for information_schema.schemata.
+type InfoSchemaSchemataExtractor struct {
+	InfoSchemaBaseExtractor
+}
+
+// ListSchemas lists related schemas from predicate.
+// If no schema found in predicate, it return all schemas.
+func (e *InfoSchemaSchemataExtractor) ListSchemas(is infoschema.InfoSchema) []model.CIStr {
+	return e.listSchemas(is, _schemaName)
+}
+
+func (e *InfoSchemaBaseExtractor) listSchemas(is infoschema.InfoSchema, schemaCol string) []model.CIStr {
+	schemas := e.getSchemaObjectNames(schemaCol)
+	if len(schemas) == 0 {
+		ret := is.AllSchemaNames()
+		slices.SortFunc(ret, func(a, b model.CIStr) int {
+			return strings.Compare(a.L, b.L)
+		})
+		return ret
+	}
+	ret := schemas[:0]
+	for _, s := range schemas {
+		if n, ok := is.SchemaByName(s); ok {
+			ret = append(ret, n.Name)
+		}
+	}
+	return ret
+}
+
+func findNameAndAppendToTableMap(
+	ctx context.Context,
+	is infoschema.InfoSchema,
+	schema model.CIStr,
+	tableNames []model.CIStr,
+	tables map[int64]*model.TableInfo,
+) error {
+	for _, n := range tableNames {
+		tbl, err := is.TableByName(ctx, schema, n)
+		if err != nil {
+			if terror.ErrorEqual(err, infoschema.ErrTableNotExists) {
+				continue
+			}
+			return errors.Trace(err)
+		}
+		tblInfo := tbl.Meta()
+		tables[tblInfo.ID] = tblInfo
+	}
+	return nil
+}
+
+// findTablesByID finds tables by table IDs and append them to table map.
+func findTablesByID(
+	is infoschema.InfoSchema,
+	tableIDs []model.CIStr,
+	tableNames []model.CIStr,
+	tables map[int64]*model.TableInfo,
+) {
+	for _, tid := range parseIDs(tableIDs) {
+		tbl, ok := is.TableByID(tid)
+		if !ok {
+			continue
+		}
+		if len(tableNames) > 0 && containInTableNames(tableNames, tbl) {
```

Comment:
maybe use map type for `tableNames`, to implement O(1) `containInTableNames`.

Reply:
https://github.com/pingcap/tidb/pull/55144/commits/05c4815e2e440b952991717f395e1b56e1bba8c0